src/python/pants/backend/python/rules/coverage.py

```diff
@@ -0,0 +1,279 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import configparser
+import json
+from dataclasses import dataclass
+from io import StringIO
+from typing import Dict, Tuple
+import itertools
+
+import pkg_resources
+
+from pants.backend.python.rules.inject_init import InjectedInitDigest
+from pants.backend.python.rules.pex import (
+  CreatePex,
+  Pex,
+  PexInterpreterConstraints,
+  PexRequirements,
+)
+from pants.backend.python.rules.prepare_chrooted_python_sources import ChrootedPythonSources
+
+from pants.backend.python.rules.python_test_runner import (
+  DEFAULT_COVERAGE_CONFIG,
+  get_coveragerc_input,
+)
+from pants.backend.python.subsystems.pytest import PyTest
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.backend.python.subsystems.python_tool_base import PythonToolBase
+from pants.backend.python.subsystems.subprocess_environment import SubprocessEncodingEnvironment
+from pants.base.specs import AddressSpecs
+from pants.build_graph.address import Address
+from pants.engine.addressable import BuildFileAddresses
+from pants.engine.fs import (
+  Digest,
+  DirectoriesToMerge,
+  DirectoryWithPrefixToAdd,
+  FileContent,
+  FilesContent,
+  InputFilesContent,
+)
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.legacy.graph import HydratedTarget, TransitiveHydratedTargets, HydratedTargets
+from pants.engine.rules import goal_rule, rule, subsystem_rule
+from pants.engine.selectors import Get, MultiGet
+from pants.rules.core.strip_source_root import SourceRootStrippedSources
+from pants.rules.core.test import AddressAndTestResults, PytestCoverageReport
+from pants.source.source_root import SourceRootConfig
+
+
+class CoverageToolBase(PythonToolBase):
+  options_scope = 'pytest-coverage'
+  default_version = 'coverage==5.0.0'
+  default_entry_point = 'coverage'
+  default_interpreter_constraints = ["CPython>=3.6"]
+
+
+  @classmethod
+  def register_options(cls, register):
+    super().register_options(register)
+    register(
+      '--output-path',
+      type=str,
+      default='coverage/python',
+      help='Path to write pytest coverage report to. Must be relative to build root.',
+    )
+    register(
+      '--report',
+      type=str,
+      choices=('xml', 'html'),
+      default='html',
+      help='Which coverage reports to emit.',
+    )
+
+
+@dataclass(frozen=True)
+class CoverageSetup:
+  requirements_pex: Pex
+  filename: str
+
+
+@rule
+async def setup_coverage(
+  coverage: CoverageToolBase,
+) -> CoverageSetup:
+  plugin_file_digest: Digest = await Get[Digest](InputFilesContent, get_coverage_plugin_input())
```

Comment:
Could you please double check if you can remove the `Digest` on the left? MyPy should be able to correctly infer the type from the `await Get`.